client/consensus/slots/src/lib.rs

```diff
@@ -475,7 +475,7 @@ pub enum CheckedHeader<H, S> {
 
 #[derive(Debug, thiserror::Error)]
 #[allow(missing_docs)]
-pub enum Error<T> where T: SlotData + Clone + Debug + Send + Sync + 'static {
+pub enum Error<T> where T: SlotData + Debug + Send + Sync + 'static {
```

Comment:
```suggestion
pub enum Error<T> {
```

Please tell me again why we can not remove this?^^

Reply:
The minimum bound is `T: Debug`, since that is used for the display impl of `SlotDuration<T>`.

But the whole error type becomes pointless and unusable if it's not `Send + Sync + 'static` - it can't be used as `source()` of other error types.

I am aware that the impl already requires `impl<T: Clone + Debug + Send + Sync + 'static> SlotDuration<T> { ..` but this would still allow using `Error` type without the constraints which would be non intuitive.

Reply:
I can repeat myself again, if a bound is not required for a certain type declaration, it should not be added. And here we don't need it. We may require it at some point, but than we need to add the bounds there.

And clearly, `Debug' is here also not required :P

Reply:
On a general note: https://rust-lang.github.io/api-guidelines/future-proofing.html?#data-structures-do-not-duplicate-derived-trait-bounds-c-struct-bounds.

Reply:
`#[error("{:?}")]` requires `Debug` for `T`, since otherwise `SlotDuration<T>` does not fulfill - https://github.com/paritytech/substrate/pull/7631/files#diff-2ff27f67b9a711481f3fe7f514fd50c9f63a2bd2852fd47610dbd1a4c758b061R478-R479 
The derived implementation depends on `T` implementing `Debug`. Removal leads to a compiler error. Ways forward are to impl `Debug` for `SlotDuration<T>` manually, regardless of `T` constraints, or add the `Debug` constraint. If there is a third way, I'd be happy to be enlightened.

Reply:
@ordian that clarifies things a bit, the automatic derived impl with `#[error("{0:?})]` does not handle add the required trait bounds to the generics.

Reply:
> On a general note: https://rust-lang.github.io/api-guidelines/future-proofing.html?#data-structures-do-not-duplicate-derived-trait-bounds-c-struct-bounds.

Yeah that is what I meant :D

Reply:
> `#[error("{0:?}")]` requires `Debug` for `T`, since otherwise `SlotDuration<T>` does not fullfil `Debug` - https://github.com/paritytech/substrate/pull/7631/files#diff-2ff27f67b9a711481f3fe7f514fd50c9f63a2bd2852fd47610dbd1a4c758b061R478-R479
> The derived implementation depends on `T` implementing `Debug`. Removal leads to a compiler error. Ways forward are to impl `Debug` for `SlotDuration<T>` manually, regardless of `T` constraints, or add the `Debug` constraint. If there is a third way, I'd be happy to be enlightened.

Okay that looks like missing support from thiserror.

Reply:
In this case we actually need `T: Debug` to generate `impl Display for Error<T>`, which is required for `impl std::error::Error for Error<T>`.

Reply:
Filed https://github.com/dtolnay/thiserror/issues/112 upstream.